handler/datasource/prepare.go

```diff
@@ -0,0 +1,296 @@
+	}
+}
+
+func (frs *fileRangeSet) addIfFits(fileRange model.FileRange, maxSize int64) bool {
+	nextSize := toCarSize(fileRange.Length)
+	if frs.carSize+nextSize > maxSize {
+		return false
+	}
+	frs.fileRanges = append(frs.fileRanges, fileRange)
+	frs.carSize += nextSize
+	return true
+}
+
+func (frs *fileRangeSet) reset() {
+	frs.fileRanges = make([]model.FileRange, 0)
+	frs.carSize = carHeaderSize
+}
+
+func (frs *fileRangeSet) fileRangeIDs() []uint64 {
+	return underscore.Map(frs.fileRanges, func(fileRange model.FileRange) uint64 {
+		return fileRange.ID
+	})
+}
+
+func toCarSize(size int64) int64 {
+	out := size
+	nBlocks := size / 1024 / 1024
+	if size%(1024*1024) != 0 {
+		nBlocks++
+	}
+
+	// For each block, we need to add the bytes for the CID as well as varint
+	out += nBlocks * (36 + 9)
+
+	// For every 256 blocks, we need to add another block.
+	// The block stores up to 256 CIDs and integers, estimate it to be 12kb
+	if nBlocks > 1 {
+		out += (((nBlocks - 1) / 256) + 1) * 12000
+	}
+
+	return out
+}
+
+func nextAvailablePackJob(
+	ctx context.Context,
+	db *gorm.DB,
+	sourceID uint32,
+) (*model.PackJob, error) {
+	var packJob model.PackJob
+	err := database.DoRetry(ctx, func() error {
+		return db.Where(model.PackJob{SourceID: sourceID, PackingState: model.Created}).Preload("FileRanges").FirstOrCreate(&packJob).Error
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &packJob, nil
+}
+
+func PrepareSource(ctx context.Context, db *gorm.DB, datasourceHandlerResolver datasource.HandlerResolver, source model.Source, scanSource bool) error {
+	directoryCache := make(map[string]uint64)
+	dataset := *source.Dataset
+	var remainingParts []model.FileRange
+	err := db.Joins("File").
+		Where("source_id = ? AND file_ranges.pack_job_id is null", source.ID).
+		Order("file_ranges.id asc").
+		Find(&remainingParts).Error
+	if err != nil {
+		return err
+	}
+	logger.With("remaining", len(remainingParts)).Info("remaining files")
+
+	if scanSource {
+		sourceScanner, err := datasourceHandlerResolver.Resolve(ctx, source)
+		if err != nil {
+			return fmt.Errorf("failed to get source scanner: %w", err)
+		}
+		entryChan := sourceScanner.Scan(ctx, "", source.LastScannedPath)
+		for entry := range entryChan {
+			if entry.Error != nil {
+				logger.Errorw("failed to scan", "error", entry.Error)
+				continue
+			}
+			file, fileRanges, err := PushFile(ctx, db, entry.Info, source, dataset, directoryCache)
+			if err != nil {
+				return fmt.Errorf("failed to push file: %w", err)
+			}
+			if file == nil {
+				logger.Infow("file already exists", "path", entry.Info.Remote())
+				continue
+			}
+			err = database.DoRetry(ctx, func() error {
+				return db.Model(&model.Source{}).Where("id = ?", source.ID).
+					Update("last_scanned_path", file.Path).Error
+			})
+			if err != nil {
+				return fmt.Errorf("failed to update last scanned path: %w", database.ErrDatabaseNotSupported)
+			}
+
+			remainingParts = append(remainingParts, fileRanges...)
+		}
```

Comment:
@hannahhoward 
Here are the moved code I'm talking about. 
Before the change - packJobs are created while the source is being scanned
After the change - packJobs are created after the source scanning completes